### PR TITLE
prefix indexes

### PR DESCRIPTION
**Pull request by staltz**
For issue #40, this is a work in progress but most of it is working.

- [x] Experiments with 8 bit, 16 bit, 32 bit
- [x] Working implementation
- [x] Benchmarks
- [x] Consider the API
- [x] Tests
- [x] Clean up code

Note about the API: I think for a long time what would be a good operator name. But it turns out that semantically this is the same as `equal`, it's just a different way of doing `equal`, so I opted for the same name, but indicating that it's using prefix indexes `equalViaPrefix`. I would maybe prefer another API: `equal(seeker, value, opts)` where `opts: {indexType, indexAll, prefix}` so to use this new algorithm you would write `equal(seekVoteLink, msgid, {prefix:true})` or `slowEqual('value.content.vote.link', msgid, {prefix:true})`

Random notes about early perf benchmarks:

```
and(blog, or(author1, author2))

  - 3 records found
  - duration when cold everything missing: 6900ms
  - duration when cold 3 missing indexes:  3000ms
  - duration when cold 2 missing indexes:  2100ms
  - duration when cold 1 missing index:    2000ms
  - duration when warm query:                20ms

votes.link=msgid (prefix 16 bits from base64 decoded)

  - 24 records found
  - duration when cold everything missing:   5300ms
  - duration when cold prefix index missing: 4000ms
  - duration when warm query:                  50ms

slowEqualViaPrefix votes.link=msgid (prefix 32 bits from base64 as is in log.bipf)

  - 17 records found, some are false positives
  - duration when cold everything missing:   4800ms
  - duration when cold prefix index missing: 3300ms
  - duration when warm query:                  40ms

equalViaPrefix votes.link=msgid (prefix 32 bits from base64 as is in log.bipf)

  - 17 records found, some are false positives
  - duration when cold everything missing:   3600ms
  - duration when cold prefix index missing: 2000ms
  - duration when warm query:                  30ms

equalViaPrefix votes.link=msgid (prefix 32 bits from base64 as is in log.bipf)
with additional iteration over the results to remove false positives

  - 15 records found
  - duration when cold everything missing:   4600ms
  - duration when cold prefix index missing: 2100ms
  - duration when warm query:                  40ms
```

The prefix index file is the same size as the `offset.index` and `sequence.index`, roughly ~5MB for a production dataset.

### CODE REVIEW

**Comment by arj03** [comment]
First of all great work on this. I like the concept overall. It's pretty crazy that its so quick to loop over a million values, modern processors are really optimized for loops :-)

As for API, I like this one: `equal(seeker, value, opts)`

For db2 indexes, I could see this used for vote + root links in social. But I guess not mentions? And for base it would be interesting to try this for key index as that is probably quite big in base. And this is going to really interesting to test for about and friends.

I have similar query numbers:

Vote query: 35ms
Vote query db2: 26ms

I'll do a review of the code now.

File: index.js [hunk 1]
```diff
@@ -133,17 +145,29 @@ module.exports = function (log, indexesPath) {
...
+  function savePrefixIndex(name, prefixIndex, count, cb) {
+    debug('saving prefix index: %s', name)
+    const num = prefixIndex.prefix
+    const filename = path.join(indexesPath, name + `.${num}prefix`)
+    saveTypedArrayFile(filename, prefixIndex.seq, count, prefixIndex.tarr, cb)
+  }
+
+  function growTarrIndex(index, Type) {
+    let newLength = index.tarr.length * 2
+    // FIXME: even better would be to bring back the exponential jumps, but
```
Comment (arj03) [inline index.js:157]: Yes, we really need to exponential back here. There is quite a big perf diff. Maybe with a trim at save function.

File: index.js [hunk 2]
```diff
@@ -199,11 +223,26 @@ module.exports = function (log, indexesPath) {
...
+      index.seq = seq
+      index.count = offset + 1
+      return true
+    }
+  }
+
   function updateIndexValue(opData, index, buffer, offset) {
     const status = checkValue(opData, buffer)
     if (status) index.bitset.add(offset)
-
```
Comment (arj03) [inline index.js:205]: Nice, even better make this function a 1 line

File: index.js [hunk 3]
```diff
@@ -274,20 +316,31 @@ module.exports = function (log, indexesPath) {
...
+          else saveIndex(op.data.indexName, index)
+        }
 
         cb()
       },
     })
   }
 
-  function createIndexes(missingIndexes, cb) {
+  function createIndexes(orphanOps, cb) {
```
Comment (arj03) [inline index.js:329]: I actually like the missingIndexes name better. I'm not exactly sure what orphan means here.

Comment (staltz) [reply index.js:329]: Oh, let me explain (and I'm open to changing it back): I general take the name to indicate the type of the thing, and in this case `missingIndexes` was not an array of indexes, it was an array of `op.data`. This confused me for a while, and changing the name helped me to know that these were actually ops, and they were passed to other functions like `updateIndexValue(opData, .....)`

**Comment by staltz** [comment]
> First of all great work on this. I like the concept overall.

Thanks!

> I have similar query numbers:
>
> Vote query: 35ms
> Vote query db2: 26ms

Those are using leveldb? Do you think that 40ms with prefix indexes is acceptable or not acceptable?

> For db2 indexes, I could see this used for vote + root links in social. But I guess not mentions? And for base it would be interesting to try this for key index as that is probably quite big in base. And this is going to really interesting to test for about and friends.

Yeah I'm not sure how to do mentions, we could perhaps support the first mention in the mentions array, but not all of them, unless we do a prefix index for the first mention, another prefix index for the second mention, the third mention, etc, but that seems wasteful.

**Comment by arj03** [comment]
> > I have similar query numbers:
> > Vote query: 35ms
> > Vote query db2: 26ms
> 
> Those are using leveldb? Do you think that 40ms with prefix indexes is acceptable or not acceptable?

I think we should weight this for the different scenarios. It might be possible to make this a bit faster as well.

> > For db2 indexes, I could see this used for vote + root links in social. But I guess not mentions? And for base it would be interesting to try this for key index as that is probably quite big in base. And this is going to really interesting to test for about and friends.
> 
> Yeah I'm not sure how to do mentions, we could perhaps support the first mention in the mentions array, but not all of them, unless we do a prefix index for the first mention, another prefix index for the second mention, the third mention, etc, but that seems wasteful.

I have a hunch that we might make a compound prefix for these that would have less resolution but in practice might work out. Lets iterate on that idea after this. It's not really related, was just thinking ahead :-)

File: index.js (hunk 3 above)
Comment (arj03) [reply index.js:329]: How about opsNotIndexed instead then?

File: index.js (hunk 1 above)
Comment (staltz) [reply index.js:157]: So, I'm not sure what to do here, but I'll share some numbers:

- exponential jumps unlimited
  - `and(blog, or(author1, author2))` with every index missing: 4400ms
  - `and(blog, or(author1, author2))` with non-core indexes missing: 3000ms
  - `and(blog, or(author1, author2))` warm query: 18ms
  - `equal(seekVoteLink, msgid, prefix)` with every index missing: 3400ms
  - `equal(seekVoteLink, msgid, prefix)` with non-core indexes missing: 2000ms
  - `equal(seekVoteLink, msgid, prefix)` warm query: 45ms
- exponential jumps limited to 64000
  - `and(blog, or(author1, author2))` with every index missing: 4600ms
  - `and(blog, or(author1, author2))` with non-core indexes missing: 3000ms
  - `and(blog, or(author1, author2))` warm query: 18ms
  - `equal(seekVoteLink, msgid, prefix)` with every index missing: 3500ms
  - `equal(seekVoteLink, msgid, prefix)` with non-core indexes missing: 2100ms
  - `equal(seekVoteLink, msgid, prefix)` warm query: 40ms

So there's a small CPU perf difference, but more significant storage difference.

**Comment by staltz** [comment]
> It might be possible to make this a bit faster as well.

Yep, I was excited to let you review this and see if there's anything smarter I can do. 

Also, I forgot now what the leveldb vote index in ssb-db2 looked like in performance for the cold case, where it needs to build the leveldb before answering the query. Do you remember roughly what it was?

> I have a hunch that we might make a compound prefix for these that would have less resolution but in practice might work out. Lets iterate on that idea after this. It's not really related, was just thinking ahead :-)

:100: for "low resolution" strategies

**Comment by staltz** [comment]
Idea: compression of the prefix index files might make a huge difference (I suppose primarily in storage? Maybe CPU diference too, although I'm not sure how)

**Comment by arj03** [comment]
> Yep, I was excited to let you review this and see if there's anything smarter I can do.

I'll give it a go a bit later today or tomorrow. It shouldn't change anything here and I think we could even merge this without.
 
> Also, I forgot now what the leveldb vote index in ssb-db2 looked like in performance for the cold case, where it needs to build the leveldb before answering the query. Do you remember roughly what it was?

social db2 index on my machine is roughly 13 seconds.

**Comment by staltz** [comment]
> social db2 index on my machine is roughly 13 seconds.

oh that's a lot, it's a big gain if we can take it down from 13 seconds cold to 2 seconds cold.

**Comment by arj03** [comment]
Sure but that is for 3 indexes: vote, root and mentions. But yes, if we can get it down to 6 then 👍

**Comment by arj03** [comment]
Base is actually the worst. Its like double of social.

**Comment by staltz** [comment]
> Sure but that is for 3 indexes: vote, root and mentions. But yes, if we can get it down to 6 then +1

Oh right, it's one leveldb for all those. hmm...

**Comment by staltz** [comment]
Idea: put a cache in `getRecord` and make also `getMessage` use getRecord's cache. Might be smart because `matchAgaintPrefix` anyway has to fetch the record to check if it's a false positive, and then again to deliver the results it does another fetch.

**Comment by staltz** [comment]
@arj03 Check the last commit, I managed to shave off some milliseconds by avoiding Promises (they are based on the microscheduler which may delay things a tiny bit) and by avoiding extra for loops.

Now we're down to roughly **28ms**. Note that the image below also shows Profiler-specific functions:

Screenshot from 2020-12-03 13-55-28

File: index.js (hunk 1 above)
Comment (staltz) [reply index.js:157]: So I think I'll put unlimited exponential jumps back, and for the storage solution, compression. What do u think?

File: index.js (hunk 3 above)
Comment (staltz) [reply index.js:328]: Even better

File: index.js (hunk 1 above)
Comment (arj03) [reply index.js]: There is a count in saveCoreIndex that could be used to just trim of the 0s at the end I think on save. As for compression maybe we can do that in a separate PR?

**Comment by staltz** [comment]
ready for a final review!

**Comment by staltz** [comment]
oh, one last task would be updating README

**Comment by staltz** [comment]
updated readme

**Comment by arj03** [comment]
I reviewed the code. Looks good. There is one idea I want to test before we merge and close this.

Superb work :rocket:

**Comment by staltz** [comment]
I can do the tarr index file cut off of empty space in a separate PR

**Comment by arj03** [comment]
Here is the thing that I wanted to test. Instead of looping over all the results to find all the values, sort the prefixes and then do a binary search on that. The idea is that when you load the vote index you can sort it once and then all queries for vote links will be faster. It seems that this + the optimization above means that we might even be faster than level :-)

Example code:

```
    const bitset = new TypedFastBitSet()
    const target = op.data.value
    const targetPrefix = target.readUInt32LE(0)
    const prefixIndex = indexes[op.data.indexName]

    console.time("sort")
    let t = {}
    for (let i = 0; i < prefixIndex.count; ++i) {
      let r = t[prefixIndex.tarr[i]] || []
      r.push(i)
      t[prefixIndex.tarr[i]] = r
    }

    let sortedKeys = Object.keys(t).sort((x, y) => x - y)
    console.timeEnd("sort")

    console.time("sort find")
    const start = bsb.ge(sortedKeys, targetPrefix)
    const end = bsb.gt(sortedKeys, targetPrefix)
    let res = sortedKeys.slice(start, end)
    res.forEach(x => bitset.add(x))
    console.log("results sorted", t[res])
    console.timeEnd("sort find")
  
    console.time("loopy")
    for (let i = 0; i < prefixIndex.count; ++i) {
      if (prefixIndex.tarr[i] === targetPrefix) bitset.add(i)
    }
    console.timeEnd("loopy")
    console.log("results loopy", bitset.array())
```
and output:

```
sort: 481.513ms
results sorted [ 308836, 504010, 611882, 916688 ]
sort find: 1.899ms
loopy: 13.774ms
results loopy [ 308836, 504010, 611882, 916688 ]
```

**Comment by arj03** [comment]
Merging, we can do these kind of optimizations as separate PRs

**Comment by staltz** [comment]
Very cool idea to sort it! can you make a release of this so I can test in db2? probably you will also test, but i want to test too because i want to have some fun

**Comment by arj03** [comment]
Ha, yes but won't have time until later tonight. 

0.22 released. Will also add you to npm tonight so you can do release of jitdb and ssb-db2.

**Pull request merged**